### DISCUSSION TITLE
`System.AccessViolationException` When `Bitmap.DecodeToWidth`

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#13239] [Q&A]
```
try
{
    var data1 = await BaseClient.GetBytes(url);
    if (data1.Item1)
    {
        using var stream = new MemoryStream(data1.Item2!);
        var image = Bitmap.DecodeToWidth(stream, 80);
        image.Save(Local + sha1);
        return image;
    }

    return null;
}
catch (Exception e)
{
    Logs.Error(App.GetLanguage("Gui.Error24"), e);
}
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/d5b42f38-204d-430f-b47d-370c1dd6f897)

**Comment by maxkatz6** [comment]
Avalonia passes input stream directly to the SkiaSharp: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Skia/Avalonia.Skia/ImmutableBitmap.cs#L65C39-L65C54
It might make sense to raise an issue there, depending on the content of the stream.

It's a common problem, when `stream.Position` is in the wrong position, but shouldn't be an issue here.

**Comment by Coloryr** [comment]
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/1539d98e-2d71-4859-b9a0-513d7739e2b5)
It may image problem? or SkiaSharp problem?

use `ImageSharp` to read image work well.

**Comment by Coloryr** [comment]
Is look like `SkiaSharp` problem.

https://github.com/mono/SkiaSharp/issues/2645

https://github.com/AvaloniaUI/Avalonia/blob/1ee3779a7ab61fab4090c9178f2f2e298d87671a/build/SkiaSharp.props#L3C1-L3C1

**Comment by agnauck** [comment]
have the same issue with certain images.
Looks like we have to wait for the Skia fix

**Reply by thevortexcloud** [reply]
I don't know if I have had this exact problem but I have also had `DecodeToWidth` crash with certain images (seemingly only jpegs). I am however not sure what the problem was since all it did was give me a vague exit code with no exceptions (which is the main reason I have not reported it). I ended up working around it by just using Skia to open the image, scale the image myself and then convert it to a working image format, and then load that into an Avalonia bitmap. Which works fine. EG

```csharp
              using (var fs = File.Open(ImagePath, FileMode.Open, FileAccess.Read, FileShare.Read)) {
                    using (var skbit = SKBitmap.Decode(fs)) {
                        // Now scale that to the size that we want
                        var realScale = skbit.Height < skbit.Width ? (double)skbit.Height / skbit.Width : (double)skbit.Width / skbit.Height;
                       //Scale to a max width of 200
                        using (var resized = skbit.Resize(new SKSizeI(200, (int)(realScale * 200)), SKFilterQuality.Medium)) {
                            using (var enc = resized.Encode(SKEncodedImageFormat.Webp, 10)) {
                                using (var ms = new MemoryStream()) {
                                    enc.SaveTo(ms);
                                    ms.Position = 0;
                                    //var bitmap = await Task.Run(() => Bitmap.DecodeToHeight(fs, 100));
                                    //Thumbnail = new Bitmap(ms);
                                    return new Bitmap(ms);
                                }
                            }
                        }
                    }
                }
```

**Comment by agnauck** [comment]
Hello @thevortexcloud,

yes its also jpg images that crash for me. Its crashing the whole app and the exceptions cannot be catched and ignored.
The failing images can also not be loaded with  `new Bitmap(imageStream))`. This returns null, but without crashing the whole app at least.

Thanks for sharing your code, this solves the problem and the images which fail can be loaded, resized and displayed with that solution 👍

**Reply by thevortexcloud** [reply]
You are welcome. I have found that code and variations of it can get around the majority of Skia and Avalonia image related bugs for what ever reason.

**Comment by agnauck** [comment]
Looks like its crashing here in ImmutableBitmap with some jpgs:
https://github.com/AvaloniaUI/Avalonia/blob/0a6987b0855bc7cd0b371571bb0234dc779da181/src/Skia/Avalonia.Skia/ImmutableBitmap.cs#L72

My avatar is one of them. I attached it if someone wants to debug.
[1.zip](https://github.com/AvaloniaUI/Avalonia/files/13628710/1.zip)

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/pull/13900#issuecomment-1850339413 if anyone wants to test newer SkiaSharp

**Reply by Coloryr** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/cccd23b3-6f5e-4c10-b112-5d68acd74857)
Well...

**Reply by timunie** [reply]
Do you have the nuget source for Avalonia-All installed? `https://nuget-feed-all.avaloniaui.net/v3/index.json`

**Reply by Coloryr** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/d946ddec-4ac5-40d9-ada1-2c9ed8c53ffe)
Nope.

**Reply by timunie** [reply]
thx for checking anyways :-)

**Comment by Coloryr** [comment]
link to issue #13359

**Comment by kekekeks** [comment]
Could somebody check what Skia does internally when decoding to specified size? If it does allocate a temporary bitmap for unscaled image, we should probably do the same manually as a workaround?

**Reply by timunie** [reply]
I'm trying to create a failing test. Let me check if that helps debugging.

**Reply by thevortexcloud** [reply]
I _think_ it does this.

https://github.com/google/skia/blob/b7ac9da3e5ac34baf5da70b3e37cdbfd898ed54b/src/codec/SkJpegCodec.cpp#L407-L440

The exact logic varies by the codec. However Skiasharp itself just seems to be calling some API wrapper which I can't find the code for so I am not clear if that's where it ends up.

https://github.com/mono/SkiaSharp/blob/25e70a390e2128e5a54d28795365bf9fdaa7161c/binding/SkiaSharp/SKCodec.cs#L42-L47

**Reply by timunie** [reply]
```
    [DllImport("libSkiaSharp", CallingConvention = (CallingConvention) 2)]
    internal static extern unsafe void sk_codec_get_scaled_dimensions(
      IntPtr codec,
      float desiredScale,
      SKSizeI* dimensions);
```
 Out of my head. Will file the Test anyway, so if someone wants to take the branch would be nice.

**Comment by perfect100** [comment]
So, the tutorial of [Music Store App](https://github.com/AvaloniaUI/Avalonia/discussions/13239) can not work. Do we need to give a caution here? 

```C#
public async Task LoadCover()
{
    await using (var imageStream = await _album.LoadCoverBitmapAsync())
    {
        Cover = await Task.Run(() => Bitmap.DecodeToWidth(imageStream, 400));
    }
}
```

**Reply by thevortexcloud** [reply]
It can work. You just need to be selective about what images you use as only some will cause the issue. Or just use the work around I posted earlier. But yeah, adding some sort of warning that image scaling via `Bitmap.DecodeToWidth` is currently not working properly for some image types, and linking back to this discussion is probably a good idea.

**Reply by timunie** [reply]
I've mixed feelings here. I really hate that we cannot work around that bug in SkiaSharp atm. We need to wait for them to fix it. At the same time we cannot downgrade for a good reason. Adding a warning to the docs may stay forever which is also not perfect but probably worth it now if someone would want to add it.

**Reply by thevortexcloud** [reply]
I assume the XML docs in Avalonia itself should probably be updated as well to make it more obvious? Although I am not sure if that is a good idea since `Bitmap` is just a wrapper around what ever the implementation is. But it's one of the most likely places people will see it, which will probably prevent more people from opening issues.

I will also point out, this is not the first time I have run into a Skia bug like this. For a fairly long time loading most PNGs was also broken (and was also what prevented me from upgrading to a new version of Avalonia for a long time). From memory it took about 8 months to fix. So if that is any indication this is probably going to be around for awhile.

**Comment by TCFox** [comment]
I encountered this issue too (while making the Music Store App tutorial), Avalonia 11.0.6 seems to use a buggy version of SkiaSharp which results in this issue further downstream somewhere.

Simply updating to 11.0.9 worked for me. If you encounter this issue, try updating all NuGet Packages to their latest patched version as this may resolve the issue.

**Reply by thevortexcloud** [reply]
It has indeed been officially fixed in newer versions of Avalonia.
